### ISSUE TITLE
Error with query filters

### ISSUE THREAD

**Issue by david-garcia-garcia** [zzzprojects/EntityFramework-Extensions#168]
Hi, we are seeing random failures in our IC servers due to the following that looks some sort of race condition or concurrency issue:

Error message:
System.NullReferenceException : Object reference not set to an instance of an object.
Error stack trace:
   at Z.EntityFramework.Plus.QueryFilterContextInterceptor.Initialize(DbContext context)
   at Z.EntityFramework.Plus.QueryFilterManager.AddOrGetFilterContext(DbContext context)
   at Z.EntityFramework.Plus.QueryFilterInterceptorDbCommandTree.TreeCreated(DbCommandTreeInterceptionContext interceptionContext)
   at System.Data.Entity.Infrastructure.Interception.InternalDispatcher`1.Dispatch[TInterceptionContext,TResult](TResult result, TInterceptionContext interceptionContext, Action`2 intercept)
   at System.Data.Entity.Infrastructure.Interception.DbCommandTreeDispatcher.Created(DbCommandTree commandTree, DbInterceptionContext interceptionContext)
   at System.Data.Entity.Core.Common.DbProviderServices.CreateCommandDefinition(DbCommandTree commandTree, DbInterceptionContext interceptionContext)
   at System.Data.Entity.Core.Objects.Internal.ObjectQueryExecutionPlanFactory.CreateCommandDefinition(ObjectContext context, DbQueryCommandTree tree)
   at System.Data.Entity.Core.Objects.Internal.ObjectQueryExecutionPlanFactory.Prepare(ObjectContext context, DbQueryCommandTree tree, Type elementType, MergeOption mergeOption, Boolean streaming, Span span, IEnumerable`1 compiledQueryParameters, AliasGenerator aliasGenerator)
   at System.Data.Entity.Core.Objects.ELinq.ELinqQueryState.GetExecutionPlan(Nullable`1 forMergeOption)
   at System.Data.Entity.Core.Objects.ObjectQuery`1.<>c__DisplayClass7.<GetResults>b__6()
   at System.Data.Entity.Core.Objects.ObjectContext.ExecuteInTransaction[T](Func`1 func, IDbExecutionStrategy executionStrategy, Boolean startLocalTransaction, Boolean releaseConnectionOnSuccess)
   at System.Data.Entity.Core.Objects.ObjectQuery`1.<>c__DisplayClass7.<GetResults>b__5()
   at System.Data.Entity.SqlServer.DefaultSqlExecutionStrategy.Execute[TResult](Func`1 operation)
   at System.Data.Entity.Core.Objects.ObjectQuery`1.GetResults(Nullable`1 forMergeOption)
   at System.Data.Entity.Core.Objects.ObjectQuery`1.<System.Collections.Generic.IEnumerable<T>.GetEnumerator>b__0()
   at System.Data.Entity.Internal.LazyEnumerator`1.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)

**Comment by JonathanMagnan** [comment]
Hello @david-garcia-garcia ,

Could you provide an example of how your filter is created? (info@zzzprojects.com if you need to keep the file private)

That usually happens when the filter is created at the wrong place (for example, adding global filter on every context)

Best Regards,

Jonathan

---

**Performance Libraries**
`context.BulkInsert(list, options => options.BatchSize = 1000);  `
[Entity Framework Extensions](http://entityframework-extensions.net/) • [Entity Framework Classic](http://entityframework-classic.net/) • [Bulk Operations](http://bulk-operations.net/) •  [Dapper Plus](http://dapper-plus.net/) •  [LinqToSql Plus](http://linqtosql-plus.net/)

**Runtime Evaluation**
`Eval.Execute("x + y", new {x = 1, y = 2}); // return 3`
[C# Eval Function](http://eval-expression.net/) • [SQL Eval Function](http://eval-sql.net/)

**Comment by david-garcia-garcia** [comment]
I'll try to post here the relevant code and explain.

The filters are added to every instance of a DataContext:

```
        protected void InitializeFiltersForContext(DbContext currentContext, CompiledModelStore model)
        {
               [...]

                foreach (var filter in entityFilters.Value)
                {
                    this.AddFilter(currentContext, entityFilters.Key, filter.Key, filter.Value);
                }
            }
        }

        protected void AddFilter(DbContext currentContext, Type entity, string key, object filter, bool enabled = true)
        {
            // Invoke: public static BaseQueryFilterInterceptor Filter<T>(this DbContext context, object key, Func<IQueryable<T>, IQueryable<T>> queryFilter, bool isEnabled = true) where T : class;
            MethodInfo addFilterMethod = typeof(QueryInterceptorFilterExtensions).GetMethods()[2];
            MethodInfo addFilterGenericMethod = addFilterMethod.MakeGenericMethod(entity);
            addFilterGenericMethod.Invoke(currentContext, new object[] { currentContext, $"{entity.FullName}-{key}", filter, enabled });
        }
```
This issue only happens on test environment where each test runs on a separate thread, with concurrency.

When each test finishes, they clear ZZZ and EF caches like this:

       ```
        public static void ClearZzzAndEfCaches(string connectionString)
        {
            lock (Locker)
            {
                using (SqlConnection connection = new SqlConnection(connectionString))
                {
                    using (var ctx = new DbContext(connection, false))
                    {
                        ctx.RefreshModel();
                    }
                }

                ClearEfCaches();
            }
        }

       public static void ClearEfCaches()
        {
            MetadataWorkspace.ClearCache();
            InformationSchemaManager.ClearInformationSchemaTable();
            EntityFrameworkManager.Cache = new StaticObjectCache();
            QueryCacheManager.Cache = new StaticObjectCache();

            ClearEfInternalCaches();
        }

        public static void ClearEfInternalCaches()
        {
            var assembly = typeof(DbContext).Assembly;

            var typeLazyInternalContext = assembly.GetType("System.Data.Entity.Internal.LazyInternalContext");

            IDictionary initializedDatabases = (IDictionary)typeLazyInternalContext.GetField("InitializedDatabases", BindingFlags.NonPublic | BindingFlags.Static)?.GetValue(null);
            initializedDatabases?.Clear();

            IDictionary cachedModels = (IDictionary)typeLazyInternalContext.GetField("_cachedModels", BindingFlags.NonPublic | BindingFlags.Static)?.GetValue(null);
            cachedModels?.Clear();
       }
```

Initial exploration of the internals of ZZZ points to the possibility of a race condition when calling:

`ctx.RefreshModel();`

while a scope is being instantiated.

**Comment by david-garcia-garcia** [comment]
Any updates? Thanks!

**Comment by JonathanMagnan** [comment]
Hello @david-garcia-garcia ,

We released the v1.8.20 of Z.EntityFramework.Plus.EF6

In this version, we added some additional null check that could cause the `Object reference not set to an instance of an object.` issue.

On our side, when we compare with the stack trace, we don't think the `ctx.RefreshModel();` cause the issue but perhaps that's two different issues.

Something that looks weird on our side, is your line `MetadataWorkspace.ClearCache();`

We indeed use this namespace in this feature:

```csharp
var entityTypes = context.GetObjectContext().MetadataWorkspace.GetItemCollection(DataSpace.OSpace)
                .Where(x => x.BuiltInTypeKind == BuiltInTypeKind.EntityType);
```

Perhaps too much stuff is cleared and the MetadataWorkspace contains now some null value?

Best Regards,

Jonathan
